modcc/printer/cudaprinter.cpp

```diff
@@ -252,4 +362,106 @@ void emit_common_defs(std::ostream& out, const Module& module_) {
     out << popindent << "};\n\n";
 }
 
+void emit_api_body_cu(std::ostream& out, APIMethod* e, bool is_point_proc) {
+    auto body = e->body();
+    auto indexed_vars = indexed_locals(e->scope());
+
+    if (!body->statements().empty()) {
+        out << "int tid_ = threadIdx.x + blockDim.x*blockIdx.x;\n";
+        out << "if (tid_<n_) {\n" << indent;
+        out << "auto gid_ __attribute__((unused)) = params_.node_index_[tid_];\n";
+        out << "auto cid_ __attribute__((unused)) = params_.vec_ci_[gid_];\n";
+
+        auto uses_ion = [&] (ionKind k) {
+            for(auto &symbol : e->scope()->locals()) {
+                if (k == symbol.second->is_local_variable()->ion_channel()) {
+                    return true;
+                }
+            }
+            return false;
+        };
+
+        uses_ion(ionKind::K)  && out << "auto kid_  = params_.ion_k_index_[tid_];\n";
+        uses_ion(ionKind::Ca) && out << "auto caid_ = params_.ion_ca_index_[tid_];\n";
+        uses_ion(ionKind::Na) && out << "auto naid_ = params_.ion_na_index_[tid_];\n";
+
+        for (auto& sym: indexed_vars) {
+            emit_state_read_cu(out, sym);
+        }
+
+        out << cuprint(body);
 
+        for (auto& sym: indexed_vars) {
+            emit_state_update_cu(out, sym, sym->external_variable(), is_point_proc);
+        }
+        out << popindent << "}\n";
+    }
+}
+
+void emit_procedure_body_cu(std::ostream& out, ProcedureExpression* e) {
+    out << cuprint(e->body());
+}
+
+void emit_state_read_cu(std::ostream& out, LocalVariable* local) {
+    out << "value_type " << cuprint(local) << " = ";
+
+    if (local->is_read()) {
+        out << cuprint(local->external_variable()) << ";\n";
+    }
+    else {
+        out << "0;\n";
+    }
+}
+
+void emit_state_update_cu(std::ostream& out, Symbol* from,
+                          IndexedVariable* external, bool is_point_proc)
+{
+    if (!external->is_write()) return;
+
+    const bool is_minus = external->op()==tok::minus;
+    if (is_point_proc) {
+        out << "arb::gpu::reduce_by_key(";
+        is_minus && out << "-";
+        out << from->name()
+            << ", params_." << decode_indexed_variable(external).data_var << ", gid_);\n";
+    }
+    else {
+        out << cuprint(external) << (is_minus? " -= ": " += ") << from->name() << ";\n";
+    }
+}
+
+// Gives the name of the local variable used to index into a field.
+const char* index_id(Symbol *s) {
+    if(s->is_variable()) return "tid_";
+    else if(auto var = s->is_indexed_variable()) {
+        switch(var->ion_channel()) {
+            case ionKind::none: return "gid_";
+            case ionKind::Ca:   return "caid_";
+            case ionKind::Na:   return "naid_";
+            case ionKind::K:    return "kid_";
+            default :
+                throw compiler_exception("CudaPrinter unknown ion type", s->location());
+        }
+    }
+    return "";
+}
```

Comment:
Related to comment above about using the hard-coded ionKinds, `decode_indexed_variable` provides a variable name for the index as well as the data. This would provide names that are common with the cpu printer, and also will make arbitrary ions easier to implement.

Reply:
See comment above.